packages/odyssey-react-mui/src/test-selectors/querySelector.ts

```diff
@@ -11,188 +11,238 @@
+         */
+        role: Role[number];
+      }
+    : object
+  : object) &
+  (LocalFeatureTestSelector extends TestSelector
+    ? {
+        /**
+         * Helps narrow down HTML selection to the correct element.
+         */
+        options: Record<
+          keyof LocalFeatureTestSelector["selector"]["options"],
+          string | RegExp
+        >;
+      }
+    : object);
+
+export const querySelector =
+  <LocalFeatureTestSelector extends FeatureTestSelector>(
+    /**
+     * Selectors object including features and accessible text selections.
+     */
+    featureTestSelector: LocalFeatureTestSelector,
+  ) =>
+  <LocalQueryMethod extends QueryMethod = "get">(
+    props: {
+      /**
+       * Refers to Testing Library's canvas. This is usually `screen`, but Storybook uses `within(canvas)`.
+       */
+      element: HTMLElement;
+    } & InnerQuerySelectorProps<LocalFeatureTestSelector, LocalQueryMethod>,
+  ) => {
+    const { element: containerElement, queryMethod } = props;
+    const localQueryMethod = queryMethod || ("get" as const);
+    const querySelectorOptions = "options" in props ? props.options : undefined;
+    const role = "role" in props ? (props.role as AriaRole) : undefined;
+
+    // This `let` is difficult to make into a `const`. It makes the code unreadable.
+    let capturedElement: HTMLElement | null = null;
+
+    if ("selector" in featureTestSelector && querySelectorOptions) {
+      const sharedProps = {
+        element: containerElement,
+        queryMethod: localQueryMethod,
+        queryOptions: Object.fromEntries(
+          Object.entries(featureTestSelector.selector.options).map(
+            ([testSelectorsKey, testingLibraryKey]) => [
+              testingLibraryKey,
+              querySelectorOptions[testSelectorsKey],
+            ],
+          ),
+        ) as Record<
+          LocalFeatureTestSelector extends TestSelector
+            ? LocalFeatureTestSelector["selector"]["options"][keyof LocalFeatureTestSelector["selector"]["options"]]
+            : string,
+          string | RegExp
+        >,
+      };
+
+      capturedElement =
+        featureTestSelector.selector.method === "ByRole"
+          ? getByRoleQuerySelector({
+              ...sharedProps,
+              role:
+                Array.isArray(featureTestSelector.selector.role) || role
+                  ? role || ""
+                  : featureTestSelector.selector.role,
+            })
+          : getByTextQuerySelector({
+              ...sharedProps,
+              selectionMethod: featureTestSelector.selector.method,
+              text: featureTestSelector.selector.text,
+            });
+    } else if (
+      "isControlledElement" in featureTestSelector &&
+      featureTestSelector.isControlledElement
+    ) {
+      try {
+        capturedElement = getControlledElement({ element: containerElement });
+      } catch (error) {
+        if (queryMethod === "query") {
```

Comment:
Whats this branch about? If we're throwing immediately after it, why does it matter what this local variable is set to?

Reply:
Great catch! I meant to remove this code. It was something I was testing before I figured out the way to fix the `HTMLElement | null` issue.